---
title: "Quality Control of sc samples"
author: "Adam Cribbs, updated by Carla Cohen"
date: "`r Sys.Date()`"
output: html_document
---

# QC-1 workflow

This workflow follows the scflow quantnuclei pipeline and is part of the seurat qc-1 pipeline.

The main steps are as follows:  
* import counts matrix  
* create a Seurat Object with crude filtering  
* save RDS files
* generate and plot QC metrics  

```{r setup, include=FALSE}

library(yaml)
library(tidyverse)
library(Seurat)
library(BUSpaRse)
library(biomaRt)
library(cowplot)
library(scater)


# Read in the yml file
ini <- read_yaml("pipeline.yml")

```

###  Read in the data and create the Seurat Object


```{r, echo=FALSE, warning=FALSE}
#make a list of sample files
file_list <- list.files("kallisto.dir/", pattern="genes.mtx", recursive = TRUE)

#Make a new list for the seurat objects
so <- list()

# generate a list of seurat objects
for (i in 1:length(file_list)){
    
    name <- str_replace(file_list[i], "genes.mtx", "")
    
    res_mat <- read_count_output(dir = paste0("kallisto.dir/",name),
                                 name = "genes", 
                                 tcc = FALSE)
    
    project_title <- str_replace(file_list[i], "/bus/genecount/genes.mtx", "") %>% 
        str_remove("-H-date[:digit:].+")
    
    seurat_object <- CreateSeuratObject(res_mat, 
                             min.cells=20, 
                             min.features=ini$so_filter,
                             project = project_title
    )

    so[[i]] <- seurat_object
    
    print(project_title)
    print(so[[i]])
}

```



### Generate additional QC metrics 

* Mitochondrial reads
* number of UMIs per gene for each cell

Identify EnsemblIDs for mitochondrial genes

```{r, echo=FALSE}

# Access the useMart database
# useMart keeps not working every so often (issue https://github.com/grimbough/biomaRt/issues/31)
ensembl <- try(useMart("ensembl", dataset = "hsapiens_gene_ensembl"))
if(class(ensembl) == "try-error"){
  httr::set_config(httr::config(ssl_verifypeer = FALSE))
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
}

# get a df of ensemble_id vs gene symbol
mapping <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"), mart = ensembl)

#identify mitochondrial genes starting MT-
mito <- mapping %>% filter(grepl("^MT-", hgnc_symbol))

```

Add details of reads from mitochondrial genes, and number of UMIs per gene for each cell to  the metadata

```{r, echo=FALSE}

for(i in 1:length(file_list)){
    
    #get positions of mitochondrial genes in the so rownames
    positions <- match(mito$ensembl_gene_id, rownames(so[[i]]))
    
    #use the positions to get the list of mito genes
    mito_subset <- rownames(so[[i]])[positions]
    
    #remove NAs
    mito_subset <- mito_subset[!is.na(mito_subset)]
    
    #add percent_mt to metadata
    so[[i]][["percent_mt"]] <- PercentageFeatureSet(so[[i]], features = mito_subset)
    
    #add number of UMIs per gene for each cell to metadata
    so[[i]][["log10GenesPerUMI"]] <- log10(so[[i]][["nFeature_RNA"]]) / log10(so[[i]][["nCount_RNA"]])
}


```
### Add gene symbols to the meta.features of the RNA assay

Where no gene symbol is available, leave in the EnsemblID  
For the Seurat object this has to be added for each assay.  
For Bioconductor SingleCellExperiment this can be added as rowData.  


```{r, echo=FALSE, message = FALSE}

df <- list()
for(i in 1:length(file_list)){
    
    #make a df with ensemble_id and hgnc_symbol
    df[[i]] <- data.frame("ensembl_gene_id" = rownames(so[[i]])) %>% 
        left_join(mapping)
    
    #add a column of unique names, remove the "_" from ensembl_ids
    df[[i]]$unique_name <- uniquifyFeatureNames(df[[i]]$ensembl_gene_id, df[[i]]$hgnc_symbol)
    df[[i]]$unique_name <- gsub("^_", "", df[[i]]$unique_name)
    df[[i]] <- df[[i]] %>% distinct(ensembl_gene_id) #remove duplicated rows
    
    #add ensembl_IDs to row names
    rownames(df[[i]]) <- rownames(so[[i]])
    
    #add to the meta.features in the RNA assay
    so[[i]][["RNA"]]@meta.features <- df[[i]]
    #print(head(so[[i]][["RNA"]]@meta.features))
    
}

```

### Save the RDS objects

Save as a Seurat object.   
Convert to a SingleCellExperiment and save.  

```{r, echo=FALSE}

dir.create("RDS_objects.dir", showWarnings = FALSE)
se <- list()

for(i in 1:length(file_list)){
    
    #save the Seurat Object
    saveRDS(so[[i]], file = paste0("RDS_objects.dir/", so[[i]]@project.name, "_unfiltered_SeuratObject.rds"))
    
    #Create a single cell object
    se[[i]] <- as.SingleCellExperiment(so[[i]])
    
    # add the unique gene IDs to the rowData
    rowData(se[[i]]) <- so[[i]]@assays$RNA@meta.features
    
    # Save the single cell object
    saveRDS(se[[i]], paste0("RDS_objects.dir/", so[[i]]@project.name, "_raw_unfiltered_SingleCellExp.rds"))
}


```



## Assessing the quality metrics

We will explore the following metrics through visualizations to decide on which cells are low quality and should be removed from the analysis:

* Cell counts
* UMI counts per cell
* Genes detected per cell
* UMIs vs. genes detected
* Mitochondrial counts ratio
* Novelty

Each metric is plotted and individual plots are saved in QC_Fitures.dir.

#### Cell Counts

Plot the total number of cells per sample

```{r, echo=FALSE}

dir.create ("QC_Figures.dir", showWarnings = FALSE)

#create a df of number of cells per sample
df <- data.frame()

for (i in 1:length(file_list)) {
    sample <- so[[i]]@project.name
    cells <- ncol(so[[i]])
    df[i,1] <- sample
    df[i, 2] <- cells
}

colnames(df) <- c("sample", "cells")

#plot the number of cells per sample as a bar chart
p <- ggplot(df, aes(x = sample, y = cells, fill = sample))+
    geom_col()+
    ggtitle("Number of cells per sample")+
    theme(axis.text.x = element_text(angle = 45, hjust=1))+
    theme(legend.position="none")

p

```


```{r, include=FALSE, fig.height=10}
#save the plot
ggsave("QC_Figures.dir/nCells.png", p, device = "png")
```


#### UMI counts per cell

Plot the number of counts per cell (nCount_RNA)

The UMI counts per cell should generally be above 500, although usable, itâ€™s still low if between 500-1000 counts. If UMIs per cell is 500-1000 counts, then the cells probably should have been sequenced more deeply.

```{r, echo=FALSE, fig.height=10}

plot_list <- list()

for (i in 1:length(file_list)) {
    
    # do a violin plot
    plot_list[[i]] <- VlnPlot(so[[i]], "nCount_RNA", pt.size = 0.01)+
        #remove text on x axis
        theme(axis.text.x = element_blank())+
        #add a title
        ggtitle(so[[i]]@project.name) +
        theme(plot.title = element_text(size = 12))+
        #remove the legend
        theme(legend.position="none")
}

#add an overall title and arrange the graphs on one page
title <- ggdraw() + draw_label("ncount_RNA", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 24, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p


```
```{r, include=FALSE}
#save the plot
ggsave("QC_Figures.dir/ncount_RNA.png", p, device = "png")


```

#### Genes detected per cell

Plot the number of genes per cell (nFeature_RNA)

Seeing gene detection in the range of 500-5000 is normal for inDrop analysis. Similar expectations for gene detection as for UMI detection, although may be a bit lower than UMIs.

```{r, echo=FALSE, fig.height=10}

plot_list <- list()
for (i in 1:length(file_list)) {
  plot_list[[i]] <- VlnPlot(so[[i]], "nFeature_RNA", pt.size = 0.01)+
      theme(axis.text.x = element_blank())+
      ggtitle(so[[i]]@project.name) +
      theme(plot.title = element_text(size = 12))+
      theme(legend.position="none")
}

title <- ggdraw() + draw_label("nFeature_RNA", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p


```

```{r, include=FALSE}
#save the plot
ggsave("QC_Figures.dir/nFeature_RNA.png", p, device = "png", width = 10, height = 10, bg = "white")


```


#### Mitochondrial counts ratio

Plot the mitochondrial content (percent_mt)

```{r, echo=FALSE, fig.height=10}

plot_list <- list()
for (i in 1:length(file_list)) {
  plot_list[[i]] <- VlnPlot(so[[i]], "percent_mt", pt.size = 0.01)+
      theme(axis.text.x = element_blank())+
      ggtitle(so[[i]]@project.name) +
      theme(plot.title = element_text(size = 12))+
      theme(legend.position="none")
}

title <- ggdraw() + draw_label("percent_mt", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p


```

```{r, include=FALSE}
#save the plot
ggsave("QC_Figures.dir/percent_mt.png", p, device = "png", bg = "white", width = 10, height = 10)


```


#### UMIs vs. genes detected 

Poor quality cells are likely to have low genes and UMIs per cell. Therefore, a poor sample is likely to have cells in the lower left of the graph. Good cells should exhibit both higher number of genes per cell and higher numbers of UMIs. We also expect similar lines with similar slopes for all samples.

```{r, echo=FALSE, fig.height=10}

plot_list <- list()
for (i in 1:length(file_list)) {
  plot_list[[i]] <- FeatureScatter(so[[i]], feature1 = "nCount_RNA", feature2 = "nFeature_RNA")+
      theme(axis.text.x = element_blank())+
      ggtitle(so[[i]]@project.name) +
      theme(plot.title = element_text(size = 12))+
      theme(legend.position="none")
}

title <- ggdraw() + draw_label("UMIs vs genes detected", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p


```

```{r, include=FALSE}
#save the plot
ggsave("QC_Figures.dir/UMI_vs_genes.png", p, device = "png", width = 14, height = 14, bg = "white")

```


Plot UMIs vs genes coloured by percent_mt

```{r, echo=FALSE, message=FALSE, fig.height=10, fig.width=10}

plot_list <- list()
for (i in 1:length(file_list)) {
  
    metadata <- so[[i]][[]]

    plot_list[[i]] <- metadata %>% 
        ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent_mt)) + 
        geom_point() + 
        stat_smooth(method=lm) +
        scale_x_log10() + 
        scale_y_log10() + 
        geom_vline(xintercept = 800) +
        ggtitle(so[[i]]@project.name) +
        theme(plot.title = element_text(size = 12))
}

title <- ggdraw() + draw_label("UMIs vs genes detected", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 2))

p


```

```{r, include=FALSE}
#save the plot
ggsave("QC_Figures.dir/UMI_vs_genes_mt.png", p, device = "png", bg = "white", width = 14, height = 14)


```


#### Novelty 

Visualize the overall novelty of the gene expression by visualizing the genes detected per UMI

We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturating the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric. Generally, we expect the novelty score to be above 0.80.

```{r, echo=FALSE, fig.height=10}

plot_list <- list()
for (i in 1:length(file_list)) {
  
    metadata <- so[[i]][[]]

    plot_list[[i]] <- metadata %>% 
        ggplot(aes(x=log10GenesPerUMI, fill = sample)) +
        geom_density() +
        ggtitle(so[[i]]@project.name)+
        theme(plot.title = element_text(size = 12))+
        theme(legend.position="none")
}

title <- ggdraw() + draw_label("Novelty", fontface='bold', size = 24)
p <- plot_grid(plotlist = plot_list, label_size = 12, ncol = 3)
p <- plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))

p

```

```{r, include=FALSE}
#save the plot
ggsave("QC_Figures.dir/Novelty.png", p, device = "png", bg = "white", width = 12, height = 12)


```

